File: apps/api/plane/db/models/case_cluster.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

from django.db import models
from django.db.models import Q

from .project import ProjectBaseModel


class CaseClusterAnalysisStatus(models.TextChoices):
    PROCESSING = "processing", "Processing"
    READY = "ready", "Ready"
    FAILED = "failed", "Failed"


class CaseClusterCategory(models.TextChoices):
    SCENE = "scene", "Scene"
    REQUIREMENT = "requirement", "Requirement"
    BADCASE = "badcase", "Badcase"
    OPPORTUNITY = "opportunity", "Opportunity"


class CaseClusterAnalysis(ProjectBaseModel):
    name = models.CharField(max_length=255)
    status = models.CharField(
        max_length=20,
        choices=CaseClusterAnalysisStatus.choices,
        default=CaseClusterAnalysisStatus.PROCESSING,
    )
    model_name = models.CharField(max_length=255, blank=True)
    source_count = models.PositiveIntegerField(default=0)
    error = models.TextField(blank=True)
    raw_response = models.JSONField(default=dict)

    class Meta:
        db_table = "case_cluster_analyses"
        ordering = ("-created_at",)


class CaseCluster(ProjectBaseModel):
    analysis = models.ForeignKey(
        "db.CaseClusterAnalysis",
        on_delete=models.CASCADE,
        related_name="clusters",
    )
    category = models.CharField(max_length=20, choices=CaseClusterCategory.choices)
    name = models.CharField(max_length=255)
    summary = models.TextField(blank=True)
    typical_feedback = models.TextField(blank=True)
    product_opportunity = models.TextField(blank=True)
    sort_order = models.FloatField(default=65535)
    merged_into = models.ForeignKey(
        "self",
        on_delete=models.SET_NULL,
        related_name="merged_clusters",
        null=True,
        blank=True,
    )

    class Meta:
        db_table = "case_clusters"
        ordering = ("sort_order", "created_at")


class CaseClusterItem(ProjectBaseModel):
    analysis = models.ForeignKey(
        "db.CaseClusterAnalysis",
        on_delete=models.CASCADE,
        related_name="items",
    )
    cluster = models.ForeignKey(
        "db.CaseCluster",
        on_delete=models.CASCADE,
        related_name="items",
    )
    issue = models.ForeignKey(
        "db.Issue",
        on_delete=models.PROTECT,
        related_name="case_cluster_items",
    )
    source_snapshot = models.JSONField(default=dict)
    sort_order = models.FloatField(default=65535)

    class Meta:
        db_table = "case_cluster_items"
        ordering = ("sort_order", "created_at")
        constraints = [
            models.UniqueConstraint(
                fields=["analysis", "issue"],
                condition=Q(deleted_at__isnull=True),
                name="case_cluster_item_unique_analysis_issue",
            )
        ]

File: packages/types/src/case-cluster.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TCaseClusterCategory = "scene" | "requirement" | "badcase" | "opportunity";
export type TCaseClusterAnalysisStatus = "processing" | "ready" | "failed";

export type TCaseClusterSourceSnapshot = {
  id: string;
  identifier: string;
  name: string;
  description: string;
  priority: string;
  state: string;
  type: string;
  labels: string[];
};

export type TCaseClusterItem = {
  id: string;
  issue_id: string;
  cluster_id: string;
  source_snapshot: TCaseClusterSourceSnapshot;
  sort_order: number;
  created_at: string;
  updated_at: string;
};

export type TCaseCluster = {
  id: string;
  analysis_id: string;
  category: TCaseClusterCategory;
  name: string;
  summary: string;
  typical_feedback: string;
  product_opportunity: string;
  sort_order: number;
  merged_into_id: string | null;
  items: TCaseClusterItem[];
  created_at: string;
  updated_at: string;
};

export type TCaseClusterAnalysis = {
  id: string;
  name: string;
  status: TCaseClusterAnalysisStatus;
  model_name: string;
  source_count: number;
  error: string;
  clusters: TCaseCluster[];
  created_at: string;
  updated_at: string;
};

export type TCaseClusterCandidate = TCaseClusterSourceSnapshot;

export type TCaseClusterUpdate = Pick<
  TCaseCluster,
  "category" | "name" | "summary" | "typical_feedback" | "product_opportunity"
>;
